queenbee/io/alias.py

```diff
@@ -0,0 +1,568 @@
+
+    See http://json-schema.org/understanding-json-schema/reference/array.html for
+    more information.
+    """
+    type: constr(regex='^DAGArrayInputAlias$') = 'DAGArrayInputAlias'
+
+    default: List = Field(
+        None,
+        description='Default value to use for an input if a value was not supplied.'
+    )
+
+    items_type: ItemType = Field(
+        ItemType.String,
+        description='Type of items in an array. All the items in an array must be from '
+        'the same type.'
+    )
+
+    @validator('default', always=True)
+    def replace_none_value(cls, v):
+        return [] if not v else v
+
+    def validate_spec(self, value):
+        """Validate an input value against specification.
+
+        Use this for validating workflow inputs against a recipe.
+        """
+        if self.spec:
+            spec = dict(self.spec)
+            spec['type'] = 'array'
+            spec['items'] = self.items_type.lower()
+            json_schema_validator(value, spec)
+
+
+class DAGObjectInputAlias(DAGGenericInputAlias):
+    """An alias JSON object input.
+
+    JSON objects are similar to Python dictionaries.
+
+    You can add additional validation by defining a JSONSchema specification.
+
+    See http://json-schema.org/understanding-json-schema/reference/object.html for
+    more information.
+    """
+    type: constr(regex='^DAGObjectInputAlias$') = 'DAGObjectInputAlias'
+
+    default: Dict = Field(
+        None,
+        description='Default value to use for an input if a value was not supplied.'
+    )
+
+    @validator('default', always=True)
+    def replace_none_value(cls, v):
+        return {} if not v else v
+
+    def validate_spec(self, value):
+        """Validate an input value against specification.
+
+        Use this for validating workflow inputs against a recipe.
+        """
+        if self.spec:
+            spec = dict(self.spec)
+            spec['type'] = 'object'
+            json_schema_validator(value, spec)
+
+
+DAGAliasInputs = Union[
+    DAGGenericInputAlias, DAGStringInputAlias, DAGIntegerInputAlias, DAGNumberInputAlias,
+    DAGBooleanInputAlias, DAGFolderInputAlias, DAGFileInputAlias, DAGPathInputAlias,
+    DAGArrayInputAlias, DAGObjectInputAlias
+]
+
+
+class DAGGenericOutputAlias(FromOutput):
+    """DAG generic alias output.
+
+    In most cases, you should not be using the generic output unless you need a dynamic
+    output that changes its type in different platforms because of returning different
+    objects in handler.
+    """
+    type: constr(regex='^DAGGenericOutputAlias$') = 'DAGGenericOutputAlias'
+
+    platform: List[str] = Field(
+        ...,
+        description='Name of the client platform (e.g. Grasshopper, Revit, etc). The '
+        'value can be any strings as long as it has been agreed between client-side '
+        'developer and author of the recipe.'
+    )
+
+    handler: List[IOAliasHandler] = Field(
```

Comment:
Renaming the key to handlers?

Reply:
see 👆 - I will change this based on the decision that we make up there.